Feature: Farming Status Tracker (Discord Webhook)

ILike2WatchMemes opened the pull request:
## What
Added a Farming Status Tracker that periodically sends player farming-related stats to a Discord webhook allowing users to monitor farming performance and status remotely.
This feature is useful for:
    * Tracking long-term farming sessions
    * Monitoring performance and buffs remotely
    * Keeping records or debugging optimization changes

<details>
<summary>Images</summary>

![grafik](https://github.com/user-attachments/assets/6c67ded4-9f63-4d5f-bd27-84e8c6fc0d96)

</details>

## Changelog New Features
+ Added Farming Status Tracker. - ILike2WatchMemes
    * Sends an embed message including various user-selected stats to a user-defined Discord webhook every 1–10 minutes.

## Changelog Technical Details
+ Implemented a JSON PATCH method in ApiUtil to support partial updates. - ILike2WatchMemes



ignore_from_changelog
reverted for now:
https://github.com/hannibal002/SkyHanni/commit/c0cd53fd51da9931bb125b7911e607135646da59

ILike2WatchMemes committed:
Added Farming Tracker

ILike2WatchMemes committed:
fixed display errors

ILike2WatchMemes committed:
reworks

ILike2WatchMemes committed:
Merge branch 'hannibal002:beta' into farming-tracker

Obsidianninja11 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+    var currentCrop: Crop? = null
+    var currentPlacement = 0.0
+    const val SKYHANNI_URL = "https://cdn.discordapp.com/attachments/1094190239532208228/1175194612550344784/bbfwfK0.png?" +
+        "ex=669cdd05&is=669b8b85&hm=f338a9e8536ab9679f91dd1322d2d3d11549c33fb702c2b08d14132a20d860ad&"
```
Will this expire or is it fine because of it being in the webhook?

ILike2WatchMemes replied:
I got it from #contrb-chat where it has been since 17th november 2023

Obsidianninja11 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+        "ex=669cdd05&is=669b8b85&hm=f338a9e8536ab9679f91dd1322d2d3d11549c33fb702c2b08d14132a20d860ad&"
+
+    private val patternGroup = RepoPattern.group("utils.webhook")
+    val webhookPattern by patternGroup.pattern(
```
Shouldn't this pattern be in WebhookUtils?

ILike2WatchMemes replied:
yes oops

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+        if (!isEnabled()) return
+        if (lastNotification.passedSince() < config.webhook.interval.minutes) return
+
+        if (playerFaceURL.isBlank()) playerFaceURL = "https://api.mineatar.io/face/${LorenzUtils.getPlayerUuid()}?scale=12"
```
extract into api

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+    var activeAnitaBuff = ""
+    var currentCrop: Crop? = null
+    var currentPlacement = 0.0
+    const val SKYHANNI_URL = "https://cdn.discordapp.com/attachments/1094190239532208228/1175194612550344784/bbfwfK0.png?" +
```
move into another class

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/garden/TrackingConfig.java:
```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    public enum Medal {
```
move the enum inside kotlin file

hannibal002 replied:
maybe extend existing ContestBracket class, or at least create a link to it

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+
+    private fun convertPlacement(placement: Double): Medal? {
+        val isFinnegan = MayorAPI.currentMayor == Mayor.FINNEGAN
+        val (requiredBronze, requiredSilver, requiredGold, requiredPlatinum, requiredDiamond) = if (isFinnegan) {
+            listOf(
+                BRONZE_MEDAL.requiredNormal,
+                SILVER_MEDAL.requiredNormal,
+                GOLD_MEDAL.requiredNormal,
+                PLATINUM_MEDAL.requiredNormal,
+                DIAMOND_MEDAL.requiredNormal,
+            )
+        } else {
+            listOf(
+                BRONZE_MEDAL.requiredFinnegan,
+                SILVER_MEDAL.requiredFinnegan,
+                GOLD_MEDAL.requiredFinnegan,
+                PLATINUM_MEDAL.requiredFinnegan,
+                DIAMOND_MEDAL.requiredFinnegan,
+            )
+        }
```
```suggestion
        val (requiredBronze, requiredSilver, requiredGold, requiredPlatinum, requiredDiamond) = Medal.values()
            .map {
                if (isFinnegan) it.requiredNormal else it.requiredFinnegan
            }
```

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,303 @@
+        lastTablist = tabList
+        val widgets = TabWidget.entries.filter { it.isActive }
+
+        val widgetLines = { widget: TabWidget -> widgets.find { it == widget }?.lines?.map { it.removeColor() } ?: changedLines }
```
what is going on here? WidgetUpdateEvent exist

ILike2WatchMemes committed:
changed SH url, moved pattern, added regex-tests

hannibal002 reviewed (changes requested):
just a first read.
Please add regex 101 tests to all regex patterns.

ILike2WatchMemes committed:
updates

- Moved Config elements around
- Added Thread Support
- Updated to use ContestBracket
- Added getPlayerSkin in APIUtil
- Changed a few formating things
- Changed Events used

ILike2WatchMemes committed:
Add FeatureToggle

Thunderblade73 committed:
stats api

Thunderblade73 committed:
suggestions 1

ILike2WatchMemes committed:
Merge branch 'hannibal002:beta' into farming-tracker

ILike2WatchMemes committed:
Merge pull request #1 from Thunderblade73/stats-api

Depend Stats API

ILike2WatchMemes committed:
reworks + offline status

ILike2WatchMemes committed:
Merge remote-tracking branch 'upstream/beta' into farming-tracker

ILike2WatchMemes committed:
Update SkyHanniMod.kt

ILike2WatchMemes committed:
Update ApiUtils.kt

hannibal002 commented:
@ILike2WatchMemes bump merge conflicts

ILike2WatchMemes committed:
Merge remote-tracking branch 'upstream/beta' into farming-tracker

ILike2WatchMemes committed:
updates

ILike2WatchMemes committed:
fix (?)

ILike2WatchMemes commented:
one more 🙏

ILike2WatchMemes committed:
fix detekt (please)

ILike2WatchMemes committed:
fix

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/ILike2WatchMemes/farming-tracker

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/ILike2WatchMemes/farming-tracker

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/SkyHanniMod.kt:
```diff
@@ -62,6 +63,11 @@ object SkyHanniMod {
         Runtime.getRuntime().addShutdownHook(
             Thread { configManager.saveConfig(ConfigFileType.FEATURES, "shutdown-hook") },
         )
+        if (feature.garden.tracking.tracking) {
```
idk if the feature is alr not null here, and this shutdown hook should always get started, not only when the feature is enabled: instead check inside the hook if the feature is enabled

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/garden/TrackingConfig.kt:
```diff
@@ -0,0 +1,145 @@
+
+class TrackingConfig {
+    @Expose
+    @ConfigOption(name = "Webhook Settings", desc = "")
```
explain in depth what this does. remember that the term webhook is a synomym for session stealer for some users

hannibal002 committed:
fixed build

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/utils/WebhookUtils.kt:
```diff
@@ -0,0 +1,167 @@
+    private fun patchPayload(payload: Payload, url: String): JsonObject =
+        ApiUtils.patchJSON(url, Gson().toJson(payload), "Discord Webhook Api").data.asJsonObject
+
+    fun sendMessageToWebhook(
```
unused

ILike2WatchMemes replied:
I added this incase we ever want to you use it in the future

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,334 @@
+                if (widgetLines[1].contains("Starts In:")) {
+                    tablistUpcomingContestPattern.matchAll(widgetLines) {
+                        getCropEnum(group("crop"))?.let { cropEnum ->
+                            activeAnitaBuff = cropEnum.display + cropEnum.emoji
```
this data should already exist somewhere in the garden api or contest api

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,334 @@
+                    }
+                } else {
+                    tablistActiveContestPattern.matchAll(widgetLines) {
+                        currentPlacement = group("placement").toDouble()
```
does not belong in this class. either alr exists in scoreboard/garden/contest api classes, or should belong in contest api if

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,334 @@
+                }
+            }
+
+            TabWidget.ACTIVE_EFFECTS -> tablistEffectsPattern.matchAll(widgetLines) {
```
god pot timer and cooke buff should both alr be somehwere else in sh code.

hannibal002 reviewed (changes requested):
code looks overall good, but a bunch of stuff that looks like code duplication to me. we already fetch almost all data you need somwehre else in the mod. either ues that data directly, or at least create the new data fetches and storages there. the farming tracker kt file shoud not hold any farming related data on its own. all that data will be used by some features eventually.
have not tested in game.

i dont see error handling: the user wants to know in chat that their webhook worked or didnt worked.

ILike2WatchMemes committed:
updates

ILike2WatchMemes committed:
detekt

ILike2WatchMemes commented:
@hannibal002 is it okay for that function to fail complexity?

ILike2WatchMemes committed:
updates

ILike2WatchMemes committed:
error handling

ILike2WatchMemes committed:
Merge branch 'beta' into farming-tracker

CalMWolfs committed:
Merge branch 'beta' into farming-tracker

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/ILike2WatchMemes/farming-tracker

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingTracker.kt:
```diff
@@ -0,0 +1,232 @@
+        return (red shl 16) or (green shl 8) or blue
+    }
+
+    private fun convertPlacement(placement: Double): ContestBracket? {
```
unused function

hannibal002 committed:
kotlin stuff

hannibal002 committed:
cleanup

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/garden/contest/ContestBracket.kt:
```diff
@@ -2,12 +2,17 @@ package at.hannibal2.skyhanni.features.garden.contest
+enum class ContestBracket(
+    val color: String,
+    val emoji: String,
+    val requiredNormal: Double,
```
required what?

ILike2WatchMemes replied:
the required % you need for the different brackets, e.g. for diamond 5% (2% when finnegan). Will remove since unused tho anyways

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/garden/TrackingConfig.kt:
```diff
@@ -0,0 +1,170 @@
+    @ConfigOption(name = "Tracking", desc = "Send an embed with the options you selected above to your specified webhook.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    var tracking: Boolean = true
```
is this the "enabled" setting? if so, please rename display name and variable

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/garden/TrackingConfig.kt:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    enum class Crop(val display: String, val emoji: String) {
```
can we use a map or function instead? there is no direct connection to the existing crop enum in garden api

hannibal002 committed:
cleanup

hannibal002 reviewed (changes requested):
why is this feature default enabled?

ILike2WatchMemes commented:
How about Farming Status (Tracker)?

hannibal002 commented:
that can also work yes!

ILike2WatchMemes committed:
updates

ILike2WatchMemes committed:
above -> below

ILike2WatchMemes committed:
renames

please tell me if I missed any

ILike2WatchMemes committed:
default disabled

ILike2WatchMemes committed:
Update FarmingStatusTracker.kt

hannibal002 reviewed (changes requested):
fix multiversion bugs

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/features/garden/tracking/FarmingStatusTracker.kt:
```diff
@@ -0,0 +1,225 @@
+import at.hannibal2.skyhanni.utils.SkyBlockUtils
+import at.hannibal2.skyhanni.utils.WebhookUtils
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import net.minecraftforge.fml.common.network.FMLNetworkEvent
```
dont use forge events

ILike2WatchMemes commented:
> fix multiversion bugs
![grafik](https://github.com/user-attachments/assets/39c15738-350c-4917-8344-b29c6f631bbe)
![face-with-bags-under-eyes_1fae9](https://github.com/user-attachments/assets/234de3c5-7a22-451b-9705-876244cfb8d8)

ILike2WatchMemes committed:
fix multi I think

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/ILike2WatchMemes/farming-tracker

hannibal002 reviewed (changes requested):
i have no idea how to test this feature. how about we add a button below/above the "source" button that shows a guide on how to use dc webhooks in chat, or links to a guide somewhere on the internet? even when the feature is default disabled, i doubt its useful for the average user when its not clear how to use this.

![image](https://github.com/user-attachments/assets/1e221448-22b7-4e56-a7d8-220adb3c887d)

ILike2WatchMemes committed:
updates

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/ILike2WatchMemes/farming-tracker

hannibal002 committed:
updated config
